File: agent/local_inventory.py
"""
Inventario local por perfil (sin DynamoDB).
"""
import json
import uuid
from pathlib import Path
from agent.profiles import get_profile_data_dir


def _inv_file(profile_id: str) -> Path:
    return get_profile_data_dir(profile_id) / "inventory.json"


def _load(profile_id: str) -> list:
    f = _inv_file(profile_id)
    return json.loads(f.read_text(encoding="utf-8")) if f.exists() else []


def _save(profile_id: str, items: list):
    _inv_file(profile_id).write_text(json.dumps(items, ensure_ascii=False, indent=2), encoding="utf-8")


def list_products(profile_id: str) -> list[dict]:
    return sorted(_load(profile_id), key=lambda x: x.get("name", ""))


def get_product(profile_id: str, sku: str) -> dict | None:
    return next((p for p in _load(profile_id) if p["sku"] == sku), None)


def upsert_product(profile_id: str, product: dict):
    items = _load(profile_id)
    items = [p for p in items if p["sku"] != product["sku"]]
    items.append(product)
    _save(profile_id, items)


def update_product(profile_id: str, sku: str, updates: dict) -> dict | None:
    items = _load(profile_id)
    for p in items:
        if p["sku"] == sku:
            p.update({k: v for k, v in updates.items() if v is not None})
            _save(profile_id, items)
            return p
    return None


def delete_product(profile_id: str, sku: str):
    items = [p for p in _load(profile_id) if p["sku"] != sku]
    _save(profile_id, items)


def low_stock_products(profile_id: str, threshold: int = 15) -> list[dict]:
    return [p for p in _load(profile_id) if 0 < int(p.get("stock", 0)) <= threshold]


def next_sku(profile_id: str) -> str:
    if not profile_id:
        return "PROD-0001"
    items = _load(profile_id)
    nums = [int(p["sku"][5:]) for p in items if p.get("sku", "").startswith("PROD-") and p["sku"][5:].isdigit()]
    return f"PROD-{(max(nums) + 1 if nums else 1):04d}"

File: desktop_java/src/main/java/com/mfagent/service/LocalInventory.java
package com.mfagent.service;

import com.mfagent.model.Product;

import java.nio.file.Path;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * CRUD de inventario local por perfil.
 * Equivalente a agent/local_inventory.py
 */
public class LocalInventory {

    private final Path file;

    public LocalInventory(String profileId) {
        this.file = ProfilePaths.dataDir(profileId).resolve("inventory.json");
    }

    public List<Product> list() {
        return JsonStore.readList(file, Product.class)
                .stream()
                .sorted(Comparator.comparing(p -> p.getName().toLowerCase()))
                .collect(Collectors.toList());
    }

    public Optional<Product> get(String sku) {
        return JsonStore.readList(file, Product.class)
                .stream()
                .filter(p -> sku.equals(p.getSku()))
                .findFirst();
    }

    public void upsert(Product product) {
        List<Product> items = JsonStore.readList(file, Product.class);
        items.removeIf(p -> p.getSku().equals(product.getSku()));
        items.add(product);
        JsonStore.writeList(file, items);
    }

    public void update(String sku, Product updates) {
        List<Product> items = JsonStore.readList(file, Product.class);
        for (Product p : items) {
            if (p.getSku().equals(sku)) {
                if (updates.getName() != null)        p.setName(updates.getName());
                if (updates.getDescription() != null) p.setDescription(updates.getDescription());
                if (updates.getPrice() > 0)           p.setPrice(updates.getPrice());
                if (updates.getCostPrice() >= 0)      p.setCostPrice(updates.getCostPrice());
                if (updates.getStock() >= 0)          p.setStock(updates.getStock());
                if (updates.getCategory() != null)    p.setCategory(updates.getCategory());
                if (updates.getImageUrl() != null)    p.setImageUrl(updates.getImageUrl());
                break;
            }
        }
        JsonStore.writeList(file, items);
    }

    public void updateStock(String sku, int newStock) {
        List<Product> items = JsonStore.readList(file, Product.class);
        items.stream().filter(p -> p.getSku().equals(sku))
                .findFirst().ifPresent(p -> p.setStock(newStock));
        JsonStore.writeList(file, items);
    }

    public void delete(String sku) {
        List<Product> items = JsonStore.readList(file, Product.class);
        items.removeIf(p -> p.getSku().equals(sku));
        JsonStore.writeList(file, items);
    }

    public List<Product> lowStock(int threshold) {
        return list().stream()
                .filter(p -> p.getStock() <= threshold)
                .collect(Collectors.toList());
    }

    /** Genera el siguiente SKU disponible (PROD-0001, PROD-0002, ...) */
    public String nextSku() {
        List<Product> items = JsonStore.readList(file, Product.class);
        int max = items.stream()
                .map(Product::getSku)
                .filter(s -> s != null && s.startsWith("PROD-"))
                .map(s -> s.substring(5))
                .filter(s -> s.matches("\\d+"))
                .mapToInt(Integer::parseInt)
                .max().orElse(0);
        return String.format("PROD-%04d", max + 1);
    }
}
